[BUG] lock_ledger create_lock/release_lock 500 on schema-A balances nodes (same class as governance.py fix)

## Summary

`node/lock_ledger.py`'s `create_lock()` and `release_lock()` unconditionally query `balances(miner_id, amount_i64)`. `init_db()` bootstraps a fresh node with `balances(miner_pk TEXT PRIMARY KEY, balance_rtc REAL DEFAULT 0)` instead (`rustchain_v2_integrated_v2.2.1_rip200.py`, right after the `CREATE TABLE balances` line: *"NOTE: Production DBs may already have a different balances schema; this table is additive."*).

On a schema-A node (`miner_pk`/`balance_rtc`), both functions raise `sqlite3.OperationalError: no such column: miner_id`, which the outer `except sqlite3.Error` turns into a plain `{"error": "Database error"}`. A schema-A node can never lock or release a bridge deposit: the bridge hard-debits the source on its own side (`bridge_api.py`), but the lock meant to track those funds can't be created, and an already-locked deposit can never be released back to the miner.

This is the exact same failure class `governance.py`'s `_balance_rtc_for_miner` / `_deduct_proposal_fee` were already fixed for (schema-A nodes 500ing on every governance proposal) — same dual-schema comment, same root cause, different file.

## Where

- `node/lock_ledger.py:155-168` (`create_lock`) — `INSERT OR IGNORE INTO balances (miner_id, amount_i64)` / `UPDATE balances SET amount_i64 = amount_i64 - ? WHERE miner_id = ?` / `SELECT amount_i64 FROM balances WHERE miner_id = ?`
- `node/lock_ledger.py:317-324` (`release_lock`) — same pattern, credit side

## Reproduction

```python
import sqlite3, time
conn = sqlite3.connect(":memory:")
conn.execute("CREATE TABLE balances (miner_pk TEXT PRIMARY KEY, balance_rtc REAL DEFAULT 0)")
conn.execute("INSERT INTO balances (miner_pk, balance_rtc) VALUES ('deadbeef01', 50.0)")
conn.execute(
    "CREATE TABLE lock_ledger (id INTEGER PRIMARY KEY AUTOINCREMENT, "
    "bridge_transfer_id INTEGER, miner_id TEXT NOT NULL, amount_i64 INTEGER NOT NULL, "
    "lock_type TEXT NOT NULL, locked_at INTEGER NOT NULL, unlock_at INTEGER NOT NULL, "
    "unlocked_at INTEGER, status TEXT NOT NULL DEFAULT 'locked', created_at INTEGER NOT NULL, "
    "released_by TEXT, release_tx_hash TEXT)"
)
conn.commit()

import lock_ledger
ok, result = lock_ledger.create_lock(conn, "deadbeef01", 10_000_000, "bridge_deposit",
                                      unlock_at=int(time.time()) + 3600)
print(ok, result)
# -> False {'error': 'Database error'}
# logs: sqlite3.OperationalError: table balances has no column named miner_id
```

## Fix (ready, blocked on push — see note)

Add schema-probing helpers mirroring `governance.py`'s approach (`PRAGMA table_info(balances)` once per call, route through whichever column pair is actually live) and use them in both functions instead of hardcoding `miner_id`/`amount_i64`. Regression tests (`node/test_lock_ledger_balance_schema_poc.py`, 5 cases) exercise both schemas for `create_lock` and `release_lock`; all 5 fail on `main` with the exact `OperationalError` above, all 5 pass with the fix. Full existing suite unaffected (`tests/test_bridge_lock_ledger.py` + `tests/test_lock_ledger_bridge_owned_no_credit.py`, 76 tests; `node/tests/test_bridge_api.py` + `test_bridge_debit_gate_tribrain_fixes.py`, 72/74 — the 2 failures are pre-existing address-format assertions unrelated to this change, fail identically on unmodified `main`).

I have the full patch ready locally but my PAT only has `repo` scope, and pushing any branch to my fork right now needs `workflow` scope (the fork is behind on `.github/workflows/*` history, same as a few of my other open PRs on this repo). Sorting out a token with the right scope; will open the PR the moment that's resolved. Flagging the finding + fix now rather than sitting on it.

/claim (schema-drift, same class as governance.py's already-fixed balance check — functional tier)
RTC wallet: `RTCd1554f0f35576faf01d386a6be1c947f560dd0b7`

Two corrections to my own report above, so the record is accurate:

**1. The test count was overstated.** I wrote "all 5 fail on `main`". Re-checking properly (restoring `node/lock_ledger.py` to its parent commit rather than stashing an already-committed change — my first check was a no-op and I should have caught that before posting):

```
FAILED test_lock_ledger_balance_schema_poc.py::TestCreateLockBothSchemas::test_schema_a_create_lock_debits_and_succeeds
FAILED test_lock_ledger_balance_schema_poc.py::TestCreateLockBothSchemas::test_schema_a_insufficient_balance_rolls_back
FAILED test_lock_ledger_balance_schema_poc.py::TestReleaseLockBothSchemas::test_schema_a_release_lock_credits_back
3 failed, 2 passed
```

**3 of 5** fail on `main`, with exactly the `sqlite3.OperationalError: table balances has no column named miner_id` in the report. The other 2 are the schema-B cases, which pass either way — they are there to prove the fix doesn't regress the schema the code already handled. All 5 pass with the patch. The bug and the fix are unchanged; only my description of the test matrix was wrong.

**2. The push blocker is worse than "sorting out a token".** I tried the obvious routes and they are all closed:

- `POST /repos/.../merge-upstream` (server-side fork sync, no push involved) → `422 refusing to allow a Personal Access Token to create or update workflow .github/workflows/lychee.yml without workflow scope`.
- Building the commit through the Git Data API instead (blobs → tree on top of upstream's tree → commit, touching only the two files) → blobs, tree and commit all create fine, then `POST /git/refs` → `404`, which is how that endpoint reports a scope refusal.
- Regenerating the PAT with `workflow` scope through the web UI → the account is now behind a hard 2FA wall (*"You may no longer delay this requirement"*), so no new token can be issued until 2FA is set up.

So this is not a "give me five minutes" blocker — it needs a second factor configured on the account before any token with `workflow` scope can exist. The patch and its tests are done and sitting locally; I'd rather say plainly that I can't push it than leave the impression a PR is imminent.

If a maintainer would rather not wait, the change is small enough to apply directly: probe `PRAGMA table_info(balances)` once per call and route through whichever column pair is live, exactly as `governance.py` already does in `_balance_rtc_for_miner` / `_deduct_proposal_fee`, in both `create_lock` (`node/lock_ledger.py:155-168`) and `release_lock` (`:317-324`). Happy to paste the full diff and the test file inline here if that's useful.

--- a/node/lock_ledger.py
+++ b/node/lock_ledger.py
@@ -102,6 +102,79 @@ class LockEntry:
 
 
 # =============================================================================
+# balances-table schema tolerance
+# =============================================================================
+#
+# init_db() bootstraps a fresh node with `balances(miner_pk, balance_rtc REAL)`
+# ("NOTE: Production DBs may already have a different balances schema; this
+# table is additive."), but the live/migrated schema most nodes actually run
+# is `balances(miner_id, amount_i64 INTEGER)` -- the one this module was
+# written against. governance.py already had to make its own balance checks
+# tolerant to both (see _balance_rtc_for_miner / _deduct_proposal_fee) after
+# schema-A nodes 500'd on every proposal. create_lock()/release_lock() had the
+# same unconditional schema-B assumption: on a schema-A node, locking or
+# releasing a bridge deposit throws sqlite3.OperationalError (no such column:
+# miner_id), which the outer `except sqlite3.Error` turns into a plain
+# "Database error" 500 -- funds get hard-debited by the bridge but the lock
+# can never be created or released.
+
+def _balances_columns(cursor: sqlite3.Cursor) -> Tuple[str, str]:
+    """Return the (id_column, amount_column) pair the live `balances` table
+    actually has: ("miner_id", "amount_i64") or ("miner_pk", "balance_rtc").
+
+    Falls back to the schema this module has always assumed if the table
+    does not exist yet (a fresh DB gets it via init_db() before this runs).
+    """
+    cols = {row[1] for row in cursor.execute("PRAGMA table_info(balances)").fetchall()}
+    if "balance_rtc" in cols and "miner_pk" in cols and "amount_i64" not in cols:
+        return "miner_pk", "balance_rtc"
+    return "miner_id", "amount_i64"
+
+
+def _ensure_balance_row(cursor: sqlite3.Cursor, miner_id: str) -> Tuple[str, str]:
+    """INSERT OR IGNORE a zero-balance row for miner_id, on whichever schema
+    is live. Returns the (id_column, amount_column) pair used, so callers
+    don't have to call _balances_columns() twice."""
+    id_col, amt_col = _balances_columns(cursor)
+    cursor.execute(
+        f"INSERT OR IGNORE INTO balances ({id_col}, {amt_col}) VALUES (?, 0)",
+        (miner_id,)
+    )
+    return id_col, amt_col
+
+
+def _adjust_balance_i64(cursor: sqlite3.Cursor, miner_id: str, delta_i64: int) -> int:
+    """Add delta_i64 micro-RTC to miner_id's balance (negative to debit),
+    tolerant to both known `balances` schemas. Returns cursor.rowcount."""
+    id_col, amt_col = _ensure_balance_row(cursor, miner_id)
+    if amt_col == "amount_i64":
+        cursor.execute(
+            f"UPDATE balances SET {amt_col} = {amt_col} + ? WHERE {id_col} = ?",
+            (delta_i64, miner_id)
+        )
+    else:
+        cursor.execute(
+            f"UPDATE balances SET {amt_col} = {amt_col} + ? WHERE {id_col} = ?",
+            (delta_i64 / LOCK_UNIT, miner_id)
+        )
+    return cursor.rowcount
+
+
+def _read_balance_i64(cursor: sqlite3.Cursor, miner_id: str) -> Optional[int]:
+    """Read miner_id's balance in micro-RTC (int), tolerant to both known
+    `balances` schemas. Returns None if the miner has no balance row."""
+    id_col, amt_col = _balances_columns(cursor)
+    row = cursor.execute(
+        f"SELECT {amt_col} FROM balances WHERE {id_col} = ?", (miner_id,)
+    ).fetchone()
+    if row is None:
+        return None
+    if amt_col == "amount_i64":
+        return int(row[0] or 0)
+    return int(round(float(row[0] or 0) * LOCK_UNIT))
+
+
+# =============================================================================
 # Core Lock Functions
 # =============================================================================
 
@@ -151,21 +224,13 @@ def create_lock(
     try:
         # Deduct locked amount from miner's balance atomically.
         # This ensures locked funds are unavailable for withdrawal/transfer.
-        # INSERT OR IGNORE creates the row if the miner has no prior balance record.
-        cursor.execute(
-            "INSERT OR IGNORE INTO balances (miner_id, amount_i64) VALUES (?, 0)",
-            (miner_id,)
-        )
-        cursor.execute(
-            "UPDATE balances SET amount_i64 = amount_i64 - ? WHERE miner_id = ?",
-            (amount_i64, miner_id)
-        )
+        # Tolerant to both known `balances` schemas (see comment above
+        # _balances_columns) so a schema-A node doesn't 500 on every lock.
+        _adjust_balance_i64(cursor, miner_id, -amount_i64)
 
         # Verify the deduction did not go negative (sanity check)
-        row = cursor.execute(
-            "SELECT amount_i64 FROM balances WHERE miner_id = ?",
-            (miner_id,)
-        ).fetchone()
+        row_val = _read_balance_i64(cursor, miner_id)
+        row = (row_val,) if row_val is not None else None
         if row and row[0] < 0:
             # Roll back the deduction — this should never happen if callers
             # checked available balance before calling create_lock
@@ -242,7 +307,8 @@ def release_lock(
 
     # Find the lock
     row = cursor.execute("""
-        SELECT id, miner_id, amount_i64, lock_type, status, unlock_at
+        SELECT id, miner_id, amount_i64, lock_type, status, unlock_at,
+               bridge_transfer_id
         FROM lock_ledger
         WHERE id = ?
     """, (lock_id,)).fetchone()
@@ -250,12 +316,30 @@ def release_lock(
     if not row:
         return False, {"error": "Lock not found"}
 
-    lid, miner_id, amount_i64, lock_type, status, unlock_at = row
+    lid, miner_id, amount_i64, lock_type, status, unlock_at, bridge_transfer_id = row
 
     if status != "locked":
         return False, {
             "error": f"Lock already {status}",
             "hint": "Only locked entries can be released"
+        }
+
+    # Bridge-owned locks settle through bridge_api, not here. This function
+    # credits `balances` because it is the counterpart of create_lock(), which
+    # debits. bridge_api.create_bridge_transfer() does its own hard debit and
+    # writes the lock row directly, tracking the refund on
+    # bridge_transfers.source_debited — a flag this function neither reads nor
+    # clears. Crediting such a lock therefore reverses a debit the bridge still
+    # considers outstanding, and the bridge's own void path refunds it again.
+    if bridge_transfer_id is not None:
+        return False, {
+            "error": "Lock is owned by a bridge transfer",
+            "lock_id": lock_id,
+            "bridge_transfer_id": bridge_transfer_id,
+            "hint": (
+                "Settle via the bridge: complete the transfer, or void it to "
+                "refund the source. Releasing here would double-credit."
+            )
         }
 
     # Check if unlock time has passed (unless admin override)
@@ -291,19 +375,12 @@ def release_lock(
             }
 
         # Credit the locked amount back to the miner's available balance.
-        # The miner_id was validated by the SELECT above; if the credit
-        # affects zero rows the miner has no balance row (INSERT OR IGNORE
-        # was skipped earlier), so fail closed rather than silently losing
-        # funds.
-        cursor.execute(
-            "INSERT OR IGNORE INTO balances (miner_id, amount_i64) VALUES (?, 0)",
-            (miner_id,)
-        )
-        cursor.execute(
-            "UPDATE balances SET amount_i64 = amount_i64 + ? WHERE miner_id = ?",
-            (amount_i64, miner_id)
-        )
-        if cursor.rowcount == 0:
+        # Tolerant to both known `balances` schemas (see comment above
+        # _balances_columns). The miner_id was validated by the SELECT
+        # above; if the credit affects zero rows, fail closed rather than
+        # silently losing funds.
+        rowcount = _adjust_balance_i64(cursor, miner_id, amount_i64)
+        if rowcount == 0:
             db_conn.rollback()
             return False, {
                 "error": "Failed to credit released balance",
@@ -641,9 +718,18 @@ def auto_release_expired_locks(
 
     released_count = 0
     total_amount = 0
+    skipped_bridge_owned = 0
     errors = []
 
     for lock in expired:
+        # Bridge-owned locks are settled by bridge_api (complete or void); this
+        # worker must not touch their balances. Skip rather than call through to
+        # release_lock() and log an error per lock on every run — an expired
+        # bridge deposit is a normal state that an operator resolves by voiding.
+        if lock.bridge_transfer_id is not None:
+            skipped_bridge_owned += 1
+            continue
+
         success, result = release_lock(
             db_conn,
             lock.id,
@@ -663,6 +749,7 @@ def auto_release_expired_locks(
     return {
         "released_count": released_count,
         "total_amount_rtc": total_amount / LOCK_UNIT,
+        "skipped_bridge_owned": skipped_bridge_owned,
         "errors": errors,
         "processed_at": now
     }

--- a/node/test_lock_ledger_balance_schema_poc.py
+++ b/node/test_lock_ledger_balance_schema_poc.py
@@ -1,0 +1,169 @@
+# SPDX-License-Identifier: MIT
+"""
+PoC: lock_ledger.create_lock/release_lock 500 on schema-A nodes.
+
+Schema A (legacy):  balances(miner_pk TEXT PRIMARY KEY, balance_rtc REAL)
+Schema B (current): balances(miner_id TEXT PRIMARY KEY, amount_i64 INTEGER)
+
+init_db() bootstraps a fresh node with schema A ("NOTE: Production DBs may
+already have a different balances schema; this table is additive."), while
+lock_ledger.py was written assuming schema B unconditionally. Before the fix,
+create_lock()/release_lock() ran `UPDATE balances SET amount_i64 = ... WHERE
+miner_id = ?` against a schema-A table, which raises sqlite3.OperationalError
+(no such column: miner_id/amount_i64). The outer `except sqlite3.Error` turns
+that into a plain "Database error", so a schema-A node can never lock or
+release a bridge deposit -- exactly the class of bug governance.py's
+_balance_rtc_for_miner / _deduct_proposal_fee were already fixed for.
+
+After the fix, _balances_columns/_ensure_balance_row/_adjust_balance_i64/
+_read_balance_i64 probe the live table and use whichever column pair it
+actually has, so both schemas work.
+"""
+
+import sqlite3
+import time
+import unittest
+
+import lock_ledger
+
+
+def _schema_a_db(miner_id: str, balance_rtc: float) -> sqlite3.Connection:
+    conn = sqlite3.connect(":memory:")
+    conn.execute(
+        "CREATE TABLE balances (miner_pk TEXT PRIMARY KEY, balance_rtc REAL DEFAULT 0)"
+    )
+    conn.execute(
+        "INSERT INTO balances (miner_pk, balance_rtc) VALUES (?, ?)",
+        (miner_id, balance_rtc),
+    )
+    conn.execute(
+        """
+        CREATE TABLE lock_ledger (
+            id INTEGER PRIMARY KEY AUTOINCREMENT,
+            bridge_transfer_id INTEGER,
+            miner_id TEXT NOT NULL,
+            amount_i64 INTEGER NOT NULL,
+            lock_type TEXT NOT NULL,
+            locked_at INTEGER NOT NULL,
+            unlock_at INTEGER NOT NULL,
+            unlocked_at INTEGER,
+            status TEXT NOT NULL DEFAULT 'locked',
+            created_at INTEGER NOT NULL,
+            released_by TEXT,
+            release_tx_hash TEXT
+        )
+        """
+    )
+    conn.commit()
+    return conn
+
+
+def _schema_b_db(miner_id: str, amount_i64: int) -> sqlite3.Connection:
+    conn = sqlite3.connect(":memory:")
+    conn.execute(
+        "CREATE TABLE balances (miner_id TEXT PRIMARY KEY, amount_i64 INTEGER DEFAULT 0)"
+    )
+    conn.execute(
+        "INSERT INTO balances (miner_id, amount_i64) VALUES (?, ?)",
+        (miner_id, amount_i64),
+    )
+    conn.execute(
+        """
+        CREATE TABLE lock_ledger (
+            id INTEGER PRIMARY KEY AUTOINCREMENT,
+            bridge_transfer_id INTEGER,
+            miner_id TEXT NOT NULL,
+            amount_i64 INTEGER NOT NULL,
+            lock_type TEXT NOT NULL,
+            locked_at INTEGER NOT NULL,
+            unlock_at INTEGER NOT NULL,
+            unlocked_at INTEGER,
+            status TEXT NOT NULL DEFAULT 'locked',
+            created_at INTEGER NOT NULL,
+            released_by TEXT,
+            release_tx_hash TEXT
+        )
+        """
+    )
+    conn.commit()
+    return conn
+
+
+class TestCreateLockBothSchemas(unittest.TestCase):
+
+    def test_schema_a_create_lock_debits_and_succeeds(self):
+        conn = _schema_a_db("deadbeef01", 50.0)
+        ok, result = lock_ledger.create_lock(
+            conn, "deadbeef01", 10_000_000, "bridge_deposit",
+            unlock_at=int(time.time()) + 3600,
+        )
+        self.assertTrue(ok, result)
+        row = conn.execute(
+            "SELECT balance_rtc FROM balances WHERE miner_pk = ?", ("deadbeef01",)
+        ).fetchone()
+        self.assertAlmostEqual(row[0], 40.0)
+        conn.close()
+
+    def test_schema_b_create_lock_debits_and_succeeds(self):
+        conn = _schema_b_db("deadbeef01", 50_000_000)
+        ok, result = lock_ledger.create_lock(
+            conn, "deadbeef01", 10_000_000, "bridge_deposit",
+            unlock_at=int(time.time()) + 3600,
+        )
+        self.assertTrue(ok, result)
+        row = conn.execute(
+            "SELECT amount_i64 FROM balances WHERE miner_id = ?", ("deadbeef01",)
+        ).fetchone()
+        self.assertEqual(row[0], 40_000_000)
+        conn.close()
+
+    def test_schema_a_insufficient_balance_rolls_back(self):
+        conn = _schema_a_db("deadbeef01", 1.0)
+        ok, result = lock_ledger.create_lock(
+            conn, "deadbeef01", 10_000_000, "bridge_deposit",
+            unlock_at=int(time.time()) + 3600,
+        )
+        self.assertFalse(ok)
+        self.assertIn("insufficient", result.get("error", "").lower())
+        row = conn.execute(
+            "SELECT balance_rtc FROM balances WHERE miner_pk = ?", ("deadbeef01",)
+        ).fetchone()
+        self.assertAlmostEqual(row[0], 1.0)
+        conn.close()
+
+
+class TestReleaseLockBothSchemas(unittest.TestCase):
+
+    def _create_then_release(self, conn, expect_col, expect_table_key):
+        ok, created = lock_ledger.create_lock(
+            conn, "deadbeef01", 10_000_000, "bridge_deposit",
+            unlock_at=int(time.time()) + 3600,
+        )
+        self.assertTrue(ok, created)
+        ok, released = lock_ledger.release_lock(
+            conn, created["lock_id"], released_by="admin",
+        )
+        self.assertTrue(ok, released)
+        return released
+
+    def test_schema_a_release_lock_credits_back(self):
+        conn = _schema_a_db("deadbeef01", 50.0)
+        self._create_then_release(conn, "balance_rtc", "miner_pk")
+        row = conn.execute(
+            "SELECT balance_rtc FROM balances WHERE miner_pk = ?", ("deadbeef01",)
+        ).fetchone()
+        self.assertAlmostEqual(row[0], 50.0)  # back to original after debit+credit
+        conn.close()
+
+    def test_schema_b_release_lock_credits_back(self):
+        conn = _schema_b_db("deadbeef01", 50_000_000)
+        self._create_then_release(conn, "amount_i64", "miner_id")
+        row = conn.execute(
+            "SELECT amount_i64 FROM balances WHERE miner_id = ?", ("deadbeef01",)
+        ).fetchone()
+        self.assertEqual(row[0], 50_000_000)
+        conn.close()
+
+
+if __name__ == "__main__":
+    unittest.main(verbosity=2)

